taichi/rhi/metal/metal_device.mm

```diff
@@ -830,17 +890,197 @@ DeviceCapabilityConfig collect_metal_device_caps(MTLDevice_id mtl_device) {
                                        PipelineCache *cache) noexcept {
   RHI_ASSERT(src.type == PipelineSourceType::spirv_binary);
   try {
-    *out_pipeline = MetalPipeline::create(*this, (const uint32_t *)src.data,
-                                          src.size, name);
+    *out_pipeline = MetalPipeline::create_compute_pipeline(
+        *this, (const uint32_t *)src.data, src.size, name);
   } catch (const std::exception &e) {
     return RhiResult::error;
   }
   return RhiResult::success;
 }
+
+std::unique_ptr<Pipeline> MetalDevice::create_raster_pipeline(
+    const std::vector<PipelineSourceDesc> &src,
+    const RasterParams &raster_params,
+    const std::vector<VertexInputBinding> &vertex_inputs,
+    const std::vector<VertexInputAttribute> &vertex_attrs, std::string name) {
+
+  // Geometry shaders aren't supported in Vulkan backend
+  RHI_ASSERT(src.size() == 2); // One vertex one fragment
+  RHI_ASSERT(src[0].type == PipelineSourceType::spirv_binary);
+  RHI_ASSERT(src[1].type == PipelineSourceType::spirv_binary);
+  RHI_ASSERT((src[0].stage == PipelineStageType::fragment &&
```

Comment:
This feels a bit weird. Maybe it's better to use two boolean flags (e.g. `has_fragment` & `has_vertex`) and a loop, makes it easier to read.